---
title: "Diversidades alfa y beta en datos metagenomicos de shotgun de cultivo de fresa, a diferentes niveles taxonomicos"
author: "Camila Silva"
date: "`r Sys.Date()`"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Observacion en diferentes niveles taxonomicos

```{r Packages}
library("phyloseq") 
library("ggplot2") 
library("igraph")
library("readr")
library("patchwork")
library("vegan")
library("GUniFrac")
library("kableExtra")
library("RColorBrewer")
```

### Cargado de datos originales 

```{r DataLoaded}

setwd("/home/camila/GIT/Tesis_Maestria/Data/fresa_solena/Data1")
outpath = "/home/camila/GIT/Tesis_Maestria/Analisis_Comparativo/Fresa_Solena/Results_img"

fresa_kraken <- import_biom("fresa_kraken.biom")
colnames(fresa_kraken@tax_table@.Data) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
fresa_kraken@tax_table@.Data <- substr(fresa_kraken@tax_table@.Data,4,100)
colnames(fresa_kraken@otu_table@.Data) <- substr(colnames(fresa_kraken@otu_table@.Data),1,6)
metadata_fresa <- read.csv2("/home/camila/GIT/Tesis_Maestria/Data/fresa_solena/Data1/metadata.csv",header =  FALSE, row.names = 1, sep = ",")
fresa_kraken@sam_data <- sample_data(metadata_fresa)
fresa_kraken@sam_data$Sample<-row.names(fresa_kraken@sam_data)
colnames(fresa_kraken@sam_data)<-c('Treatment','Samples')
samples_to_remove <- c("MP2079","MP2080","MP2088","MP2109","MP2137")
fresa_kraken_fil <- prune_samples(!(sample_names(fresa_kraken) %in% samples_to_remove), fresa_kraken)
percentages_fil <- transform_sample_counts(fresa_kraken_fil, function(x) x*100 / sum(x) )
percentages_df <- psmelt(percentages_fil)

```

De aqui en adelante solo trabajaremos con los datos filtrados por calidad, ya que no vale lapena continuar con los datos que contiene  muestras en ceros, ya que esto altera las visualizaciones.

Queremos explorar nuestras muestras a diferentes niveles taxonómicos específicos, empezando por una aglomeración de los datos a nivel de reino y filo, Para agrupar todas las OTU que tienen la misma taxonomía en un determinado rango taxonómico, utilizaremos la función ***tax_glom()***.

### A nivel de Kingdom

```{r glom_kingdom}
percentages_glom_kingdom <- tax_glom(percentages_fil, taxrank = 'Kingdom')
percentages_glom_kingdom@tax_table@.Data
```

En reino podemos ver que tenemos solamente dos grupos que son: Bacteria y Eucariota.

### A nivel de Phylum

```{r glom_phylum}
percentages_glom_phylum <- tax_glom(percentages_fil, taxrank = 'Phylum')
percentages_glom_phylum@tax_table@.Data
```

A nivel de filo, ya podemos ver mas variedad en nuestras muestras, teniendo como mayoria los filo pertenecientes a bacteria.

Usaremos la función de phyloseq ***psmelt()***, que fusiona objetos de phyloseq en un data.frame para manipularlos con paquetes como ggplot2 y vegan. Con esta creamos un dataframe con los porcentajes de "phylum", este nos dará los porcentajes de cada OTU, perteneciente a cada muestra, a nivel taxonomico de filo.

```{r percentages_df_phylum}
percentages_df_phylum <- psmelt(percentages_glom_phylum)
str(percentages_df_phylum)
head(percentages_df_phylum)
```

Ahora, creamos un data frame con los datos originales. Esta estructura nos ayudará a comparar la abundancia absoluta con la relativa 

```{r absolute_glom_phylum}
absolute_glom_phylum <- tax_glom(physeq = fresa_kraken_fil, taxrank = "Phylum")
absolute_df_phylum <- psmelt(absolute_glom_phylum)
str(absolute_df_phylum)
```

```{r phylum_colors_abs}
absolute_df_phylum$Phylum <- as.factor(absolute_df_phylum$Phylum)
phylum_colors_abs <- colorRampPalette(brewer.pal(8,"Dark2")) (length(levels(absolute_df_phylum$Phylum)))
```

```{r absolute_plot}
absolute_plot <- ggplot(data= absolute_df_phylum, aes(x=Sample, y=Abundance, fill=Phylum))+ 
  geom_bar(aes(), stat="identity", position="stack")+
  scale_fill_manual(values = phylum_colors_abs) +
  labs(title = "Abundance", x='Sample', y='Abundance') +
  theme(legend.key.size = unit(0.2, "cm"),
        legend.key.width = unit(0.25,"cm"),
        legend.position = "bottom",
        legend.direction = "horizontal",
        legend.title=element_text(size=8, face = "bold"),
        legend.text=element_text(size=6),
        text = element_text(size=12),
        axis.text.x = element_text(angle=90, size=5, hjust=1, vjust=0.5))

absolute_plot
```

```{r relative_plot}
percentages_df_phylum$Phylum <- as.factor(percentages_df_phylum$Phylum)
phylum_colors_rel<- colorRampPalette(brewer.pal(8,"Dark2")) (length(levels(percentages_df_phylum$Phylum)))
relative_plot <- ggplot(data=percentages_df_phylum, aes(x=Sample, y=Abundance, fill=Phylum))+ 
  geom_bar(aes(), stat="identity", position="stack")+
  scale_fill_manual(values = phylum_colors_rel) +
  labs(title = "Abundance", x='Sample', y='Abundance') +
  theme(legend.key.size = unit(0.2, "cm"),
        legend.key.width = unit(0.25,"cm"),
        legend.position = "bottom",
        legend.direction = "horizontal",
        legend.title=element_text(size=8, face = "bold"),
        legend.text=element_text(size=6),
        text = element_text(size=12),
        axis.text.x = element_text(angle=90, size=5, hjust=1, vjust=0.5))


relative_plot
```

Podemos ver algunas diferencias entre los diferentes filo, sin embargo, se nota que tenemos demasiados taxones para distinguir adecuadamente el color de cada uno, a menos que se tenga una abundancia muy grande.

Como nuestro objetivo es diferenciar entre muestras sanas y enfermas, le agregaremos una diferenciación entre la variable 'Treatment', con color blaco las muestras sanas y negras las muestras enfermas. 


```{r Treatment}
percentages_df_phylum$Phylum <- as.factor(percentages_df_phylum$Phylum)
phylum_colors_rel<- colorRampPalette(brewer.pal(8,"Dark2")) (length(levels(percentages_df_phylum$Phylum)))
relative_plot <- ggplot(data=percentages_df_phylum, aes(x=Sample, y=Abundance, fill=Phylum, color=Treatment))+ 
  scale_colour_manual(values=c('white','black')) +
  geom_bar(aes(), stat="identity", position="stack")+
  scale_fill_manual(values = phylum_colors_rel) +
  labs(title = "Abundance", x='Sample', y='Abundance') +
  theme(legend.key.size = unit(0.2, "cm"),
        legend.key.width = unit(0.25,"cm"),
        legend.position = "bottom",
        legend.direction = "horizontal",
        legend.title=element_text(size=8, face = "bold"),
        legend.text=element_text(size=6),
        text = element_text(size=12),
        axis.text.x = element_text(angle=90, size=5, hjust=1, vjust=0.5))


relative_plot
```

Igualmente tenemos demasiados taxones para ver una distinción entre las muestras, más adelante crearemos subconjuntos mas pequeños para tener una mejor observación de nuestros datos.

Por otro lado, podemos usar un comando llamado "unique()" para explorar cuántos filos y reinos tenemos.

```{r unique_kingdom}
unique(fresa_kraken_fil@tax_table@.Data[,"Kingdom"])
```

```{r unique_phylum}
unique(fresa_kraken_fil@tax_table@.Data[,"Phylum"])
```
Ahora, podemos ver cuantos "Eukaryota" tenemos en "Kingdom".

```{r sum_eukaryota}
sum(fresa_kraken_fil@tax_table@.Data[,"Kingdom"] == "Eukaryota")
```
y cuantos "Bacteria"

```{r sum_bacteria}
sum(fresa_kraken_fil@tax_table@.Data[,"Kingdom"] == "Bacteria")
```

Verificando así lo que notamos anteriormente con la observación de los subconjuntos aglomerados, tenemos mayor cantidad de muestras clasificadas como bacterias, que las clasificadas como eucariota.

## Diversidad Beta

Como se menciono anteriormente el objetivo de este documento es explorar las visualizaciones de: barras, alfa diversidad y beta diversidad, para subconjuntos seleccionados de nuestros datos.

Veremos aquí solo el reino Eucariota:

```{r euk_merge}
merge_Eukaryota<-subset_taxa(fresa_kraken_fil,Kingdom=="Eukaryota")
```

sacamos  las abundancias relativas,
```{r euk_percentages}
percentages_Eukaryota <- transform_sample_counts(merge_Eukaryota, function(x) x*100 / sum(x) )
percentages_Eukaryota_df <- psmelt(percentages_Eukaryota)
```

### Beta diversidad de Eukaryota

```{r euk_ordinate}
meta_ord_Eukaryota <- ordinate(physeq = percentages_Eukaryota, method = "NMDS", distance = "bray")  
plot_ordination(physeq = percentages_Eukaryota, ordination = meta_ord_Eukaryota, color = "Treatment") +
  geom_text(mapping = aes(label = colnames(fresa_kraken_fil@otu_table@.Data)), size = 3, vjust = 1.5)
```


## Eukaryota por Phylum

```{r}
ggplot(data= percentages_Eukaryota_df, aes(x=Sample, y=Abundance, fill=Phylum, color=Treatment)) + 
  scale_colour_manual(values=c('white','black')) +
  geom_bar(aes(), stat="identity", position="stack") +
    labs(title = "Abundance", x='Sample', y='Abundance', color = 'Phylum') +
    theme(legend.key.size = unit(0.2, "cm"),
          legend.key.width = unit(0.25,"cm"),
          legend.position = "bottom",
          legend.direction = "horizontal",
          legend.title=element_text(size=8, face = "bold"),
          legend.text=element_text(size=6),
          text = element_text(size=12),
          axis.text.x = element_text(angle=90, size=5, hjust=1, vjust=0.5))
```

```{r}
merge_Eukaryota_Phylum<-tax_glom(merge_Eukaryota,taxrank="Phylum")
```
sacamos  las abundancias relativas 
```{r}
percentages_Eukaryota_Phylum <- transform_sample_counts(merge_Eukaryota_Phylum, function(x) x*100 / sum(x) )
percentages_Eukaryota_Phylum_df <- psmelt(percentages_Eukaryota_Phylum)
meta_ord_Eukaryota_Phylum <- ordinate(physeq = percentages_Eukaryota_Phylum, method = "NMDS", distance = "bray")  
plot_ordination(physeq = percentages_Eukaryota_Phylum, ordination = meta_ord_Eukaryota_Phylum, color = "Treatment") + 
  geom_text(mapping = aes(label = colnames(fresa_kraken_fil@otu_table@.Data)), size = 3, vjust = 1.5) 
```
## Eukaryota por Family
```{r}
ggplot(data= percentages_Eukaryota_df, aes(x=Sample, y=Abundance, fill=Family, color=Treatment)) + 
  scale_colour_manual(values=c('white','black')) +
  geom_bar(aes(), stat="identity", position="stack") +
    labs(title = "Abundance", x='Sample', y='Abundance', color = 'Phylum') +
    theme(legend.key.size = unit(0.2, "cm"),
          legend.key.width = unit(0.25,"cm"),
          legend.position = "bottom",
          legend.direction = "horizontal",
          legend.title=element_text(size=8, face = "bold"),
          legend.text=element_text(size=6),
          text = element_text(size=12),
          axis.text.x = element_text(angle=90, size=5, hjust=1, vjust=0.5))
```
```{r}
merge_Eukaryota_Family<-tax_glom(merge_Eukaryota,taxrank="Family")
```
sacamos  las abundancias relativas 
```{r}
percentages_Eukaryota_Family <- transform_sample_counts(merge_Eukaryota_Family, function(x) x*100 / sum(x) )
percentages_Eukaryota_Family_df <- psmelt(percentages_Eukaryota_Family)
meta_ord_Eukaryota_Family <- ordinate(physeq = percentages_Eukaryota_Family, method = "NMDS", distance = "bray")  
plot_ordination(physeq = percentages_Eukaryota_Family, ordination = meta_ord_Eukaryota_Family, color = "Treatment") + 
  geom_text(mapping = aes(label = colnames(fresa_kraken_fil@otu_table@.Data)), size = 3, vjust = 1.5)
```
## Eukaryota por Genus
```{r}
ggplot(data= percentages_Eukaryota_df, aes(x=Sample, y=Abundance, fill=Genus))+ 
         geom_bar(aes(), stat="identity", position="stack") +
  theme(text = element_text(size = 5),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
```

```{r}
merge_Eukaryota_Genus<-tax_glom(merge_Eukaryota,taxrank="Genus")
```
sacamos  las abundancias relativas 
```{r}
percentages_Eukaryota_Genus <- transform_sample_counts(merge_Eukaryota_Genus, function(x) x*100 / sum(x) )
percentages_Eukaryota_Genus_df <- psmelt(percentages_Eukaryota_Genus)
meta_ord_Eukaryota_Genus <- ordinate(physeq = percentages_Eukaryota_Genus, method = "NMDS", distance = "bray")  
plot_ordination(physeq = percentages_Eukaryota_Genus, ordination = meta_ord_Eukaryota_Genus, color = "Treatment") + 
  geom_text(mapping = aes(label = colnames(fresa_kraken_fil@otu_table@.Data)), size = 3, vjust = 1.5)
```
tomando diferentes porcentajes de abundancia
```{r}
percentages_Eukaryota_Genus_df$Genus[percentages_Eukaryota_Genus_df$Abundance < 10.0] <- "Genus < 0.5% abund." #no deja poner ningun numero diferente a 0.5%
percentages_Eukaryota_Genus_df$Genus <- as.factor(percentages_Eukaryota_Genus_df$Genus)

genus_colors_rel <- colorRampPalette(brewer.pal(8,"Dark2")) (length(levels(percentages_Eukaryota_Genus_df$Genus)))
relative_plot <- ggplot(data=percentages_Eukaryota_Genus_df, aes(x=Sample, y=Abundance, fill=Genus ,color = Treatment))+ 
  geom_bar(aes(), stat="identity", position="stack") +
  scale_fill_manual(values = phylum_colors_rel) +
  theme(text = element_text(size = 5),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

relative_plot
```

```{r}
relative_plot <- ggplot(data=percentages_Eukaryota_Genus_df, aes(x=Sample, y=Abundance, fill=Genus))+ 
  geom_bar(aes(), stat="identity", position="stack") +
  scale_fill_manual(values = phylum_colors_rel) +
  theme(text = element_text(size = 5),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
relative_plot
```

sacando la beta diversidad con aglomerado de 10%
```{r}
meta_ord_Eukaryota_Genus <- ordinate(physeq = percentages_Eukaryota_Genus, method = "NMDS", distance = "bray")  
plot_ordination(physeq = percentages_Eukaryota_Genus, ordination = meta_ord_Eukaryota_Genus, color = "Treatment") + 
  geom_text(mapping = aes(label = colnames(fresa_kraken_fil@otu_table@.Data)), size = 3, vjust = 1.5)
```


## Veremos aqui solo el reino Bacteriano
```{r}
merge_Bacteria<-subset_taxa(fresa_kraken_fil,Kingdom=="Bacteria")
```
sacamos  las abundancias relativas 
```{r}
percentages_Bacteria <- transform_sample_counts(merge_Bacteria, function(x) x*100 / sum(x) )
percentages_Bacteria_df <- psmelt(percentages_Bacteria)
```
beta diversidad de Bacteria
```{r}
meta_ord_Bacteria <- ordinate(physeq = percentages_Bacteria, method = "NMDS", distance = "bray")   
plot_ordination(physeq = percentages_Bacteria, ordination = meta_ord_Bacteria, color = "Treatment") + 
  geom_text(mapping = aes(label = colnames(fresa_kraken_fil@otu_table@.Data)), size = 3, vjust = 1.5)
```
## Bacterias por Phylum
```{r}
ggplot(data= percentages_Bacteria_df, aes(x=Sample, y=Abundance, fill=Phylum))+ 
  geom_bar(aes(), stat="identity", position="stack") +
  theme(text = element_text(size = 5),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
```
```{r}
merge_Bacteria_Phylum<-tax_glom(merge_Bacteria,taxrank="Phylum")
```
sacamos  las abundancias relativas 
```{r}
percentages_Bacteria_Phylum <- transform_sample_counts(merge_Bacteria_Phylum, function(x) x*100 / sum(x) )
percentages_Bacteria_Phylum_df <- psmelt(percentages_Bacteria_Phylum)
meta_ord_Bacteria_Phylum <- ordinate(physeq = percentages_Bacteria_Phylum, method = "NMDS", distance = "bray")  
plot_ordination(physeq = percentages_Bacteria_Phylum, ordination = meta_ord_Bacteria_Phylum, color = "Treatment") + 
  geom_text(mapping = aes(label = colnames(fresa_kraken_fil@otu_table@.Data)), size = 3, vjust = 1.5)
```
## Bacterias por Family
```{r}
ggplot(data= percentages_Bacteria_df, aes(x=Sample, y=Abundance, fill=Family))+ 
  geom_bar(aes(), stat="identity", position="stack") +
  theme(text = element_text(size = 5),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
```
```{r}
merge_Bacteria_Family<-tax_glom(merge_Bacteria,taxrank="Family")
```
## sacamos  las abundancias relativas 
```{r}
percentages_Bacteria_Family <- transform_sample_counts(merge_Bacteria_Family, function(x) x*100 / sum(x) )
percentages_Bacteria_Family_df <- psmelt(percentages_Bacteria_Family)
meta_ord_Bacteria_Family <- ordinate(physeq = percentages_Bacteria_Family, method = "NMDS", distance = "bray")  
plot_ordination(physeq = percentages_Bacteria_Family, ordination = meta_ord_Bacteria_Family, color = "Treatment") + 
  geom_text(mapping = aes(label = colnames(fresa_kraken_fil@otu_table@.Data)), size = 3, vjust = 1.5)
```
## Bacterias por Genero
```{r}
ggplot(data= percentages_Bacteria_df, aes(x=Sample, y=Abundance, fill=Genus))+ 
  geom_bar(aes(), stat="identity", position="stack") +
  theme(text = element_text(size = 5),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
```
```{r}
merge_Bacteria_Genus<-tax_glom(merge_Bacteria,taxrank="Genus")
```
sacamos  las abundancias relativas 
```{r}
percentages_Bacteria_Genus <- transform_sample_counts(merge_Bacteria_Genus, function(x) x*100 / sum(x) )
percentages_Bacteria_Genus_df <- psmelt(percentages_Bacteria_Genus)
meta_ord_Bacteria_Genus <- ordinate(physeq = percentages_Bacteria_Genus, method = "NMDS", distance = "bray")  
plot_ordination(physeq = percentages_Bacteria_Genus, ordination = meta_ord_Bacteria_Genus, color = "Treatment") + 
  geom_text(mapping = aes(label = colnames(fresa_kraken_fil@otu_table@.Data)), size = 3, vjust = 1.5)

```

tomando diferentes porcentajes de abundancia
```{r}
percentages_Bacteria_Genus_df$Genus[percentages_Bacteria_Genus_df$Abundance < 10.0] <- "Genus < 0.5% abund." #no deja poner ningun numero diferente a 0.5%
percentages_Bacteria_Genus_df$Genus <- as.factor(percentages_Bacteria_Genus_df$Genus)

genus_colors_rel <- colorRampPalette(brewer.pal(8,"Dark2")) (length(levels(percentages_Bacteria_Genus_df$Genus)))
relative_plot <- ggplot(data=percentages_Bacteria_Genus_df, aes(x=Sample, y=Abundance, fill=Genus ,color = Treatment))+ 
  geom_bar(aes(), stat="identity", position="stack") +
  scale_fill_manual(values = genus_colors_rel) +
  theme(text = element_text(size = 5),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

relative_plot
```

```{r}
relative_plot <- ggplot(data=percentages_Bacteria_Genus_df, aes(x=Sample, y=Abundance, fill=Genus)) + 
  geom_bar(aes(), stat="identity", position="stack") +
  scale_fill_manual(values = genus_colors_rel) +
  theme(text = element_text(size = 5),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

relative_plot
```

sacando la beta diversidad con aglomerado de 10%
```{r}
meta_ord_Bacteria_Genus <- ordinate(physeq = percentages_Bacteria_Genus, method = "NMDS", distance = "bray")  
plot_ordination(physeq = percentages_Bacteria_Genus, ordination = meta_ord_Bacteria_Genus, color = "Treatment") + 
  geom_text(mapping = aes(label = colnames(fresa_kraken_fil@otu_table@.Data)), size = 3, vjust = 1.5)
```
