{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import math\n",
    "import gzip\n",
    "import random\n",
    "import numpy as np\n",
    "\n",
    "from pathlib import Path\n",
    "from random import seed\n",
    "from random import randint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "# funcion para lectura de archivos\n",
    "# ingresa un archivo multifasta y arroja una lista con cada secuencia (contig)\n",
    "dicc_contigs = {}\n",
    "def lectura_genoma(File):\n",
    "    lista = []\n",
    "    longitudes = []\n",
    "    with open(File,'r') as f:\n",
    "        lines=f.read() # lectura de cada linea del archivo\n",
    "        lines=lines.split('>') # identificador de '>'\n",
    "        lines=['>'+ x for x in lines[1:]] # lista con cada elemento que comienza con '>'\n",
    "        for x in lines:\n",
    "            x1 = x.replace(\">\",\"@\")\n",
    "            x2 = x1.replace(\"\\n\",\",\",1) # el primer '\\n' se reemplazapor una coma\n",
    "            x3 = x2.replace(\"\\n\",\"\") # los siguientes se quitan\n",
    "            lista.append(x3) # lista con un contig en cada elemento y su identificador\n",
    "        # convertir la lista en un diccionario        \n",
    "        for x in lista:\n",
    "            x = x.split(',')\n",
    "            dicc_contigs[x[0]] = x[1] \n",
    "        return(dicc_contigs)  \n",
    "    \n",
    "    \n",
    "# funcion que cuenta los contigs en cada genoma\n",
    "def cuenta_contigs(dicc_contigs):\n",
    "    long_total = 0\n",
    "    dicc_longitudes = {}\n",
    "    num_contigs = len(dicc_contigs) # cuenta cuantos contigs contiene cada genoma (cuantas llaves)\n",
    "    for key in dicc_contigs:\n",
    "        longitud = len(dicc_contigs[key])# calcula la longitud de cada contig (la longitud de las claves, para cada llave)\n",
    "        long_total +=  longitud # va sumando todas las longitudes de los contigs\n",
    "        dicc_longitudes[key] = longitud # diccionario de longitudes \n",
    "        # key es el identificador del contig \n",
    "        # longitud -> es la longitud de cada contig\n",
    "        # long_total -> lontitud total del genoma\n",
    "    return (dicc_longitudes,num_contigs,long_total)    \n",
    "\n",
    "\n",
    "# funcion para calcular la proporcion de los contigs\n",
    "def proporcion_contigs(dicc_contigs,dicc_longitudes,long_total):\n",
    "    dicc_proporciones = {}\n",
    "    for key in dicc_longitudes:\n",
    "        P = dicc_longitudes[key]/long_total\n",
    "        dicc_proporciones[key] = P\n",
    "    return dicc_proporciones\n",
    "\n",
    "\n",
    "# funcion para cortar reads, dado una pocision de inicio y una longitud de corte\n",
    "def cutout(contig,i,n_length): #fordward    (resive contig y entrega reads)\n",
    "    read = contig[i:i+n_length]\n",
    "    return(read) \n",
    "\n",
    "\n",
    "def cutout2(contig,i,n_length,inserto): #backward \n",
    "    # inserto = 400 (tamaño promedio del inserto), depende del secuenciador y lo da el usuario\n",
    "    # inserto -> tamaño verdadero de la molecula de DNA que esta en el secuanciador\n",
    "    j = i + inserto\n",
    "    # j = i+inserto (pocision de inicio de reversa)\n",
    "    # n_length = 150 (tamaño del read \"cortado\")\n",
    "    read_cortado = contig[j-n_length:j]\n",
    "    #i_read = contig[j-n_length:inserto]  #(quiero comenzar en la pocision j e ir hacia atras)\n",
    "    return read_cortado\n",
    "\n",
    "\n",
    "# funcion que calcula el reverso de un Read \n",
    "def reverso(read_cortado):\n",
    "    temp_list = list(read_cortado)\n",
    "    temp_list.reverse()\n",
    "    return ''.join(temp_list)\n",
    "\n",
    "\n",
    "def complementario(read_cortadoR):\n",
    "    comp = []\n",
    "    for i in range(len(read_cortadoR)):\n",
    "        if read_cortadoR[i] == \"A\":\n",
    "            comp.append(\"T\")\n",
    "        elif read_cortadoR[i] == \"T\":\n",
    "            comp.append(\"A\")\n",
    "        elif read_cortadoR[i] == \"G\":\n",
    "            comp.append(\"C\")\n",
    "        elif read_cortadoR[i] == \"C\":\n",
    "            comp.append(\"G\")         \n",
    "    return \"\".join(comp)\n",
    "\n",
    "\n",
    "def reverso_complementario(contig,i,n_length,inserto):\n",
    "    read_rev_com = complementario(reverso(cutout2(contig,i,n_length,inserto))) \n",
    "    return (read_rev_com) # entrega cada read reverso complementario\n",
    "\n",
    "\n",
    "# funcion para cortar los reads aleatoriamente\n",
    "# Al ingresar un conjunto de genomas, este nos arroja un metagenoma con varios reads tomados de los genomas iniciales\n",
    "# para esto se le entrega el numero de reads deseados por cada genoma y la longitud.\n",
    "def reads(dicc_contigs,dicc_longitudes,dicc_proporciones,n_length,num_reads_total):  \n",
    "    dicc_reads = {}  # diccionario de reads \n",
    "    dicc_reads_reverse = {} # diccionario de reverso reads\n",
    "    k = 0\n",
    "    kk = 0\n",
    "    inserto = 30 #400 #(tamaño promedio del inserto), depende del secuenciador y lo da el usuario\n",
    "    while k < num_reads_total: # número de reads total\n",
    "        for key in dicc_contigs:\n",
    "            #contig = random.choice(list(dicc_contigs.values())) # escoje un contig al azar\n",
    "            num_reads_contig = round(num_reads_total* dicc_proporciones[key]) # numero de reads x proporcion #proporcion de reads por contig\n",
    "            contig = dicc_contigs[key]\n",
    "            # la proporcion viene del diccionario proporciones de la funcion cuenta contigs \n",
    "            for kk in range(num_reads_contig): \n",
    "                newkey = key + '_' + str(kk) # identificador de contig, con el numero de read\n",
    "                i = randint(1,(len(contig)-n_length-inserto-1)) # toma una pocision de inicio al azar se debe tener en cuenta el inserto)\n",
    "                dicc_reads[newkey] = cutout(contig,i,n_length) # para el contig dado anteriormente, se corta desde la pocision i de lonjitud ln\n",
    "                dicc_reads_reverse[newkey] = reverso_complementario(contig,i,n_length,inserto)# reverso complementario\n",
    "                kk += 1\n",
    "            k += num_reads_contig # funcion para cortar los reads aleatoriamente\n",
    "        # se debe calcular cuantas veces va a escoger cada contig,dependiendo del tamaño del contig inicial      \n",
    "    return(dicc_reads,dicc_reads_reverse)\n",
    "\n",
    "\n",
    "# funcion de creacion archivo .fastq.gz\n",
    "def crea_fastqz(dicc_reads,dicc_reads_reverse,n_length,file_name):\n",
    "    file = gzip.open(file_name+'_R1.fastq.gz','wt')\n",
    "    for key in dicc_reads: \n",
    "        file.write(str(key))\n",
    "        file.write(str('\\n'))\n",
    "        file.write(str(dicc_reads[key]))\n",
    "        file.write(str('\\n'))\n",
    "        file.write(str('+'))\n",
    "        file.write(str('\\n'))\n",
    "        file.write(str('A'*n_length)) #calidad\n",
    "        file.write(str('\\n'))     \n",
    "    # agregar el reverso complementario  \n",
    "    file = open(file_name+'_R2.fastq.gz','wt')\n",
    "    for key in dicc_reads_reverse: \n",
    "        file.write(str(key))\n",
    "        file.write(str('\\n'))\n",
    "        file.write(str(dicc_reads_reverse[key]))\n",
    "        file.write(str('\\n'))\n",
    "        file.write(str('+'))\n",
    "        file.write(str('\\n'))\n",
    "        file.write(str('A'*n_length)) #calidad\n",
    "        file.write(str('\\n'))  \n",
    "    file.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
